spacy/pipeline/entity_linker.py

```diff
@@ -451,132 +445,127 @@ def predict(self, docs: Iterable[Doc]) -> ActivationsT:
             }
         if isinstance(docs, Doc):
             docs = [docs]
-        for doc in docs:
+
+        docs_iters = tee(docs, 2)
+
+        # Call candidate generator.
+        all_ent_cands = self.get_candidates(
+            self.kb,
+            (
+                SpanGroup(
+                    doc,
+                    spans=[
+                        ent for ent in doc.ents if ent.label_ not in self.labels_discard
+                    ],
+                )
+                for doc in docs_iters[0]
+            ),
+        )
+
+        for doc in docs_iters[1]:
             doc_ents: List[Ints1d] = []
             doc_scores: List[Floats1d] = []
-            if len(doc) == 0:
+            if len(doc) == 0 or len(doc.ents) == 0:
                 docs_scores.append(Ragged(ops.alloc1f(0), ops.alloc1i(0)))
                 docs_ents.append(Ragged(xp.zeros(0, dtype="uint64"), ops.alloc1i(0)))
                 continue
             sentences = [s for s in doc.sents]
-
-            # Loop over entities in batches.
-            for ent_idx in range(0, len(doc.ents), self.candidates_batch_size):
-                ent_batch = doc.ents[ent_idx : ent_idx + self.candidates_batch_size]
-
-                # Look up candidate entities.
-                valid_ent_idx = [
-                    idx
-                    for idx in range(len(ent_batch))
-                    if ent_batch[idx].label_ not in self.labels_discard
-                ]
-
-                batch_candidates = list(
-                    self.get_candidates_batch(
-                        self.kb,
-                        SpanGroup(doc, spans=[ent_batch[idx] for idx in valid_ent_idx]),
-                    )
-                    if self.candidates_batch_size > 1
-                    else [
-                        self.get_candidates(self.kb, ent_batch[idx])
-                        for idx in valid_ent_idx
-                    ]
+            doc_ent_cands = list(next(all_ent_cands))
+
+            # Looping over candidate entities for this doc. (TODO: rewrite)
```

Comment:
Why is this TODO statement here?

Reply:
That was in the [original code](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L475) already. IIRC it was related to the loop being inefficiently implemented.

Reply:
And is this fixed in the current implementation?

Reply:
I'm not aware what the TODO exactly referred to beyond some inefficiencies, so no.